applications/nrf_desktop/src/modules/watchdog.c

```diff
@@ -0,0 +1,175 @@
+	int wdt_channel_id;
+	struct k_delayed_work work;
+};
+static struct wdt_data_storage wdt_data;
+
+void watchdog_feed_worker(struct k_work *work_desc)
+{
+	struct wdt_data_storage *data =
+			CONTAINER_OF(work_desc, struct wdt_data_storage, work);
+
+	int err = wdt_feed(data->wdt_drv, data->wdt_channel_id);
+
+	if (err < 0) {
+		LOG_ERR("Cannot feed watchdog. Error code: %d", err);
+		module_set_state(MODULE_STATE_ERROR);
+		return;
+	}
+
+	k_delayed_work_submit(&data->work, WDT_FEED_WORKER_DELAY_MS);
+}
+
+void wdt_timeout_callback(struct device *dev, int channel_id)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(channel_id);
+	LOG_ERR("Watchdog timeout elapsed. Restarting the device!");
+}
+
+bool watchdog_timeout_install(struct wdt_data_storage *data)
+{
+	const struct wdt_timeout_cfg wdt_settings = {
+			.window = {
+					.min = 0,
+					.max = WDT_TIMEOUT_MS
+			},
+			.callback = wdt_timeout_callback,
+			.flags = WDT_FLAG_RESET_SOC
+	};
+
+	__ASSERT_NO_MSG(data != NULL);
+
+	data->wdt_channel_id = wdt_install_timeout(
+			data->wdt_drv, &wdt_settings);
+	if (data->wdt_channel_id < 0) {
+		LOG_ERR("Cannot install watchdog timer! Error code: %d",
+				data->wdt_channel_id);
+		return true;
+	}
+
+	LOG_INF("Watchdog timeout installed");
+	return false;
+}
+
+bool watchdog_start(struct wdt_data_storage *data)
+{
+	__ASSERT_NO_MSG(data != NULL);
+
+	int err = wdt_setup(data->wdt_drv, WDT_OPT_PAUSE_HALTED_BY_DBG);
+
+	if (err < 0) {
+		LOG_ERR("Cannot start watchdog! Error code: %d", err);
+		return true;
+	}
+
+	LOG_INF("Watchdog started");
+	return false;
+}
+
+bool watchdog_feed_enable(struct wdt_data_storage *data)
+{
+	__ASSERT_NO_MSG(data != NULL);
+
+	k_delayed_work_init(&data->work, watchdog_feed_worker);
+	int err = k_delayed_work_submit(&data->work, WDT_FEED_WORKER_START_IMMEDIATELY);
+
+	if (err < 0) {
+		LOG_ERR("Cannot start watchdog feed worker!"
+				" Error code: %d", err);
+		return true;
+	}
+
+	LOG_INF("Watchdog feed enabled");
+	return false;
+}
+
+bool watchdog_enable(struct wdt_data_storage *data)
+{
+	__ASSERT_NO_MSG(data != NULL);
+
+	data->wdt_drv = device_get_binding(WATCHDOG_DEV_NAME);
+	if (data->wdt_drv == NULL) {
+		LOG_ERR("Cannot bind watchdog driver");
+		return true;
+	}
+
+	bool err = watchdog_timeout_install(data);
+
+	if (err) {
+		return true;
+	}
+
+	err = watchdog_feed_enable(data);
+	if (err) {
+		return true;
+	}
+
+	err = watchdog_start(data);
+	if (err) {
+		return true;
+	}
+	return false;
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (is_module_state_event(eh)) {
+		const struct module_state_event *event =
+				cast_module_state_event(eh);
+
+		if (check_state(event, MODULE_ID(main), MODULE_STATE_READY)) {
+			static bool initialized;
+
+			__ASSERT_NO_MSG(!initialized);
+			initialized = true;
+
+			if (watchdog_enable(&wdt_data)) {
+				module_set_state(MODULE_STATE_ERROR);
+			} else {
+				module_set_state(MODULE_STATE_READY);
+			}
+		}
+	}
+
+	return false;
```

Comment:
`__ASSERT_NO_MSG(false)` here and return false in `is_module_state_event`

Reply:
added missing assert